msal_extensions/_windows.py

```diff
@@ -0,0 +1,173 @@
+        _memcpy(buffer, pb_data, cb_data)
+        _local_free(pb_data)
+        return buffer.raw
+
+
+# This code is modeled from a StackOverflow question, which can be found here:
+# https://stackoverflow.com/questions/463832/using-dpapi-with-python
+class WindowsDataProtectionAgent(object):
+    def __init__(self, **kwargs):
+        # type: (str) -> None
+        if 'entropy' in kwargs:
+            entropy = kwargs['entropy'].encode('utf-8')
+            self._entropy_buffer = ctypes.c_buffer(entropy, len(entropy))
+            self._entropy_blob = DATA_BLOB(len(entropy), self._entropy_buffer)
+        else:
+            self._entropy_buffer = None
+            self._entropy_blob = None
+
+    def protect(self, message):
+        # type: (str) -> bytes
+
+        message = message.encode('utf-8')
+        message_buffer = ctypes.c_buffer(message, len(message))
+        message_blob = DATA_BLOB(len(message), message_buffer)
+        result = DATA_BLOB()
+
+        if self._entropy_blob:
+            entropy = ctypes.byref(self._entropy_blob)
+        else:
+            entropy = None
+
+        if _crypt_protect_data(
+                ctypes.byref(message_blob),
+                u"python_data",
+                entropy,
+                None,
+                None,
+                _CRYPTPROTECT_UI_FORBIDDEN,
+                ctypes.byref(result)):
+            return result.raw()
+        return b''
+
+    def unprotect(self, cipher_text):
+        # type: (bytes) -> str
+        ct_buffer = ctypes.c_buffer(cipher_text, len(cipher_text))
+        ct_blob = DATA_BLOB(len(cipher_text), ct_buffer)
+        result = DATA_BLOB()
+
+        if self._entropy_blob:
+            entropy = ctypes.byref(self._entropy_blob)
+        else:
+            entropy = None
+
+        if _crypt_unprotect_data(
+            ctypes.byref(ct_blob),
+            None,
+            entropy,
+            None,
+            None,
+            _CRYPTPROTECT_UI_FORBIDDEN,
+            ctypes.byref(result)
+        ):
+            return result.raw().decode('utf-8')
+        return u''
+
+
+class _WindowsTokenCache(msal.SerializableTokenCache):
+    DEFAULT_CACHE_LOCATION = os.path.join(os.getenv('LOCALAPPDATA'), '.IdentityService', 'msal.cache')
+    DEFAULT_ENTROPY = ''
+
+    def __init__(self, **kwargs):
+        super(_WindowsTokenCache, self).__init__()
+
+        self._cache_location = _WindowsTokenCache.DEFAULT_CACHE_LOCATION  # type: str
+        if 'cache_location' in kwargs:
+            self._cache_location = kwargs['cache_location'] or _WindowsTokenCache.DEFAULT_CACHE_LOCATION
+        self._lock_location = self._cache_location + '.lockfile'
+
+        entropy = _WindowsTokenCache.DEFAULT_ENTROPY
+        if 'entropy' in kwargs:
+            entropy = kwargs['entropy']
+        self._dp_agent = WindowsDataProtectionAgent(entropy=entropy)
+        self._last_sync = 0  # _last_sync is a Unixtime
+
+    def _has_state_changed(self):
```

Comment:
My bad. I used the name `has_state_changed` in `SerializableTokenCache`. At a hindsight, I would probably rename it to `has_memory_state_changed` soon.
And then your function here should really be `has_file_state_changed()`. And these 2 flags are exactly the opposite meaning.

Reply:
I went ahead and renamed this method to `needs_refresh`, what do you think of that name?

Reply:
You ask for it! :-)

Per your request, I think the new name `needs_refresh` has the same mistake I made when I came up with the `has_state_changed` in the base class. They are both arguably vague in terms of direction. Should it be a need to handle 2-way sync in the future, the current name would become insufficient. That being said, the point is not to predict the future, because you can't. So, don't bother. :-)  Just one minor thing though. If my aforementioned mistake teaches me anything, it is that this `needs_refresh` behavior should largely be internal. So if you add a `_` prefix to make it `_needs_refresh`, it would be perfect because you can rename it later easily.

PS: Naming inside this WindowsTokenCache class matters, because it would become one of the major API of this extension module, even though we might have a factory for it.

Reply:
Fixed in e4132a1